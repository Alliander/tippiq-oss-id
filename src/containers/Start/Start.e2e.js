import secureRandomString from 'secure-random-string';
import { page as loginPage } from '../Login/Login.e2e';
import { waitUntilURLIsLoaded, getLocalstorageValue } from '../../../test/testHelper';
import { insertTestData, removeTestData } from '../../../api/common/seed-utils';
import protractorConfig from '../../../protractor.conf';

const page = {
  isLoaded: () => protractor.ExpectedConditions.presenceOf(element(by.id('page-start'))),
  heading: () => element(by.css('h1')),
  loginOptionButton: () => element(by.id('loginLink')),
  withoutSession: {
    emailAddress: () => element(by.css('.auto-register-form-component #email')),
    registerOptionButton: () => element(by.id('register')),
  },
  withSession: {
    emailAddress: () => element(by.id('email')),
    registerOptionButton: () => element(by.id('registerLink')),
  },
};

describe('Start container', () => {
  beforeAll(() => insertTestData());
  afterAll(() => removeTestData());

  beforeEach(() => {
    browser.get('/logout');
  });

  it('shows a heading on the start screen', () => {
    browser.get('/start');
    browser.wait(page.isLoaded(), 10000);
    expect(page.heading().isPresent()).toBe(true);
  });

  it('displays a supplied email address on the screen', () => {
    const randomEmail = `${secureRandomString()}@example.com`;
    browser.get(`/start?email=${randomEmail}`);
    browser.wait(page.isLoaded(), 10000);
    expect(page.withoutSession.emailAddress().getAttribute('value')).toEqual(randomEmail);
  });

  /*
  TODO: fix test, test no longer works:
    PlaceKeySelect will redirect to places first when you have no placekeys yet
  */
  xit('should redirect to PlaceKeySelect with the redirect_uri after the registration procedure', () => {
    const randomEmail = `${secureRandomString()}@example.com`;
    const redirectUri = 'http://testurl/';
    const placeKeyRedirectUri = `${protractorConfig.config.baseUrl}/selecteer-je-huis?redirect_uri=${redirectUri}`;

    browser.get(`/start?email=${randomEmail}&redirect_uri=${encodeURIComponent(redirectUri)}`);
    browser.wait(page.isLoaded(), 10000);
    page.withoutSession.registerOptionButton().click();

    waitUntilURLIsLoaded(browser, placeKeyRedirectUri);
    expect(browser.getCurrentUrl()).toEqual(placeKeyRedirectUri);
  });

  it('should display a login link to enable login with a different account', () => {
    const randomEmail = `${secureRandomString()}@example.com`;
    browser.get(`/start?email=${randomEmail}`);
    page.loginOptionButton().click();
    browser.wait(loginPage.isLoaded(), 10000);
    expect(loginPage.heading().isPresent()).toBe(true);
  });
});

describe('Start container logged in', () => {
  beforeAll(() => insertTestData());
  afterAll(() => removeTestData());

  describe('', () => {
    beforeEach(() => {
      const loginRedirectUri = encodeURIComponent('/start?redirect_uri=testurl');
      browser.get(`/login?redirect_uri=${loginRedirectUri}`);
      browser.wait(loginPage.isLoaded(), 10000);
      loginPage.emailField().sendKeys('start@test.com');
      loginPage.passwordField().sendKeys('8OO!gyxFR9qqB&We');
      loginPage.submit();
      browser.wait(page.isLoaded(), 10000);
    });

    it('displays the session email address on the screen', () => {
      expect(page.withSession.emailAddress().getText()).toEqual('start@test.com');
    });

    it('should display a login link to enable login with a different account', () => {
      page.loginOptionButton().click();
      browser.wait(loginPage.isLoaded(), 10000);
      expect(loginPage.heading().isPresent()).toBe(true);
    });

    it('should display a register link', () => {
      page.withSession.registerOptionButton().click();
      browser.wait(page.isLoaded(), 10000);
      expect(page.heading().isPresent()).toBe(true);
    });
  });

  describe('the register link', () => {
    beforeEach(() => {
      const loginRedirectUri = encodeURIComponent('/start?redirect_uri=testurl');
      browser.get(`/login?redirect_uri=${loginRedirectUri}`);
      browser.wait(loginPage.isLoaded(), 10000);
      loginPage.emailField().sendKeys('login@test.com');
      loginPage.passwordField().sendKeys('8OO!gyxFR9qqB&We');
      loginPage.submit();
      browser.wait(page.isLoaded(), 10000);
    });

    it('should logout the user', () => {
      page.withSession.registerOptionButton().click();
      browser.wait(page.isLoaded(), 10000);

      expect(page.heading().isPresent()).toBe(true);
      expect(getLocalstorageValue(browser, 'authToken')).toBe(null);
      expect(page.withoutSession.emailAddress().getAttribute('value')).toEqual('');
    });
  });
});
