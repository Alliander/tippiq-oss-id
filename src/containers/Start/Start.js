/**
 * Start container.
 * @module containers/Start
 */

import React, { Component, PropTypes } from 'react';
import Helmet from 'react-helmet';
import { Grid, Row, Col } from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { browserHistory } from 'react-router';

import { IDLE, PENDING, SUCCESS } from '../../constants/status';
import { getProfile, simpleRegister, checkEmailExists, logout } from '../../actions';
import {
  AutoRegisterForm,
  AutoRegisterPrefilled,
  PageTitle,
} from '../../components';
import { setQueryParams } from '../../utils/url';

/**
 * Start component class.
 * @function Start
 * @param {Object} props Component properties.
 * @param {Object} props.location Location properties.
 * @returns {string} Markup of the not found page.
 */
export class Start extends Component {

  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    getProfile: PropTypes.func.isRequired,
    simpleRegister: PropTypes.func.isRequired,
    checkEmailExists: PropTypes.func.isRequired,
    logout: PropTypes.func.isRequired,
    location: PropTypes.object.isRequired,
    profile: PropTypes.object.isRequired,
    simpleRegistration: PropTypes.object.isRequired,
    emailExists: PropTypes.object.isRequired,
    config: PropTypes.shape({
      tippiqPlacesBaseUrl: PropTypes.string,
    }),
  }

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Properties object
   * @constructs PlaceKeys
   */
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
  }


  /**
   * Component will mount
   * @method componentWillMount
   * @returns {undefined}
   */
  componentWillMount() {
    this.props.getProfile();
    this.props.checkEmailExists(this.props.location.query.email);
  }

  /**
   * Component will receive props
   * @method componentWillReceiveProps
   * @param {Object} nextProps Next properties
   * @returns {undefined}
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.simpleRegistration.status === SUCCESS) {
      const { tippiqPlacesBaseUrl } = this.props.config;
      const { redirect_uri: redirectUri = tippiqPlacesBaseUrl, clientId } =
        nextProps.location.query;
      window.location.href = setQueryParams('/selecteer-je-huis', {
        redirect_uri: redirectUri,
        clientId,
      });
    }

    if (nextProps.profile.status !== PENDING && !nextProps.profile.email &&
      nextProps.emailExists.success) {
      const {
        email,
        clientId,
        redirect_uri: redirectUri,
        failure_uri: failureUri,
      } = this.props.location.query;

      browserHistory
        .replace({
          pathname: '/login',
          search: `?redirect_uri=${redirectUri}` +
          `&failure_uri=${failureUri}` +
          `&clientId=${clientId}` +
          `&email=${email}`,
        });
    }
  }

  /**
   * Should component update
   * @method shouldComponentUpdate
   * @param {Object} nextProps Next properties
   * @returns {Boolean} flag whether to re-render this component or not
   */
  shouldComponentUpdate(nextProps) {
    return nextProps.profile.status !== PENDING && nextProps.emailExists !== PENDING;
  }

  /**
   * Handler for submit action.
   * @method handleSubmit
   * @param {Object} data Form data.
   * @returns {undefined}
   */
  handleSubmit(data) {
    this.props.simpleRegister(data.email);
  }

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    const { simpleRegistration, profile, emailExists } = this.props;
    const {
      email,
      clientId,
      redirect_uri: redirectUri,
      failure_uri: failureUri,
    } = this.props.location.query;

    const selectPlaceRedirectUrl = `/selecteer-je-huis?redirect_uri=${redirectUri}` +
      `&failure_uri=${failureUri}&clientId=${clientId}`;
    const registerUrl = `/start?redirect_uri=${redirectUri}` +
      `&failure_uri=${failureUri}&clientId=${clientId}`;
    const loginUrl = `/login?email=${email ? encodeURIComponent(email) : ''}&redirect_uri=${encodeURIComponent(registerUrl)}`;

    const registerForm = redirectUri && profile.email ?
      <AutoRegisterPrefilled
        email={profile.email}
        redirectUri={encodeURIComponent(selectPlaceRedirectUrl)}
        loginUrl={loginUrl}
        registerUrl={registerUrl}
        onRegisterClick={this.props.logout}
      /> :
      <AutoRegisterForm
        initialValues={{ email }}
        loginUrl={loginUrl}
        onSubmit={this.handleSubmit}
        errorMessage={simpleRegistration.error ? simpleRegistration.error.message : ''}
      />;

    return profile.status !== IDLE && emailExists.status !== PENDING ? (
      <div id="page-start">
        <Helmet title="Tippiq Account aanmaken" />
        <Grid>
          <Row>
            <Col xs={12}>
              <PageTitle>
                Tippiq <span className="text-primary">Account</span> aanmaken
              </PageTitle>
            </Col>
            <Col xs={12} sm={6} lg={7}>
              <p>
                Hiermee sluit je eenvoudig diensten aan op je Tippiq <span className="text-primary">Huis</span>.
                Bovendien kun je hiermee gemakkelijk de toegang tot je digitale huis beheren.
              </p>
            </Col>
            <Col xs={12} sm={6} lg={5} className="register-form">
              { registerForm }
            </Col>
          </Row>
        </Grid>
      </div>) : null;
  }
}

export default connect(
  state => ({
    profile: state.profile,
    simpleRegistration: state.simpleRegistration,
    emailExists: state.emailExists,
    config: state.appConfig,
  }), dispatch => bindActionCreators({
    getProfile,
    simpleRegister,
    checkEmailExists,
    logout,
  }, dispatch)
)(Start);
