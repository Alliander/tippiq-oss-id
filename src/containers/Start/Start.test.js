import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Start } from './Start';

test('Basic render snapshot test', () => {
  const component = shallow(<Start
    error={'error'}
    getProfile={() => {}}
    simpleRegister={() => {}}
    checkEmailExists={() => {}}
    logout={() => {}}
    location={{ query: '' }}
    profile={{
      status: '',
      email: '',
    }}
    simpleRegistration={{
      status: {
        success: '',
      },
      error: {
        message: '',
      },
    }}
    emailExists={{
      status: {
        success: '',
      },
      error: {
        message: '',
      },
    }}
  />);

  const tree = shallowToJson(component);
  expect(tree).toMatchSnapshot();
});
