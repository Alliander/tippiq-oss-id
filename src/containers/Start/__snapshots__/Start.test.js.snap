// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic render snapshot test 1`] = `
<div
  id="page-start"
>
  <HelmetWrapper
    title="Tippiq Account aanmaken"
  />
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        xs={12}
      >
        <PageTitle>
          Tippiq 
          <span
            className="text-primary"
          >
            Account
          </span>
           aanmaken
        </PageTitle>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        lg={7}
        sm={6}
        xs={12}
      >
        <p>
          Hiermee sluit je eenvoudig diensten aan op je Tippiq 
          <span
            className="text-primary"
          >
            Huis
          </span>
          . Bovendien kun je hiermee gemakkelijk de toegang tot je digitale huis beheren.
        </p>
      </Col>
      <Col
        bsClass="col"
        className="register-form"
        componentClass="div"
        lg={5}
        sm={6}
        xs={12}
      >
        <ConnectedForm
          errorMessage=""
          initialValues={
            Object {
              "email": undefined,
            }
          }
          loginUrl="/login?email=&redirect_uri=%2Fstart%3Fredirect_uri%3Dundefined%26failure_uri%3Dundefined%26clientId%3Dundefined"
          onSubmit={[Function]}
        />
      </Col>
    </Row>
  </Grid>
</div>
`;
